Project-Review-1

NitulAU opened the pull request:
Removed most of the bootstrap and implemented same structure using CSS.  Axios is used. Date object is introduced. will create a click handler for the next button after implementing pagination. I tried to bring out CSS and JS files. But I was getting recourse not found error. Iam working on it.

NitulAU committed:
Update index.html

Removed most of the bootstrap and implemented same structure using CSS.  Axios is used. Date object is introduced.
will create a click handler for the next button after implementing pagination. I tried to bring out CSS and JS files. But I was getting recourse not found error. Iam working on it.

NitulAU commented:
note: I can convert the CSS to SCSS format after resolving the bringing out CSS file issue. **On it**.

gavin-mannings-wl commented:
Method for loading the CSS/JS as separate files is provided in your other PR - **https://github.com/NitulAU/Gazette-Notice/pull/2#issuecomment-1517787181**

gavin-mannings-wl commented on index.html:
```diff
@@ -2,20 +2,22 @@
   <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
   <link href="https://fonts.googleapis.com/css2?family=Poppins&display=swap" rel="stylesheet">
   <!-- <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.0-alpha3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ENjdO4Dr2bkBIFxQpeoTz1HIcje39Wm4jDKdf19U8gI4ddQ3GYNS7NTKfAdVQSZe" crossorigin="anonymous"></script> -->
+  <script src="https://cdn.jsdelivr.net/npm/axios/dist/axios.min.js"></script>
```
When you are able to have separate JS files, it is good practice to not use public CDNs in your application, and instead manage the dependencies through a package manager to help manage the number of HTTP requests that a page makes, to reduce the risk of external CDNs going down and breaking your site, and to better manage security.

gavin-mannings-wl commented on index.html:
```diff
@@ -65,52 +82,92 @@
+        let card=`<div class="all-cards">`;
+        entry.forEach((item) => {
+          // console.log(item);
+          card += '<div class="notice-card mb-lg-5">';
```
Would you want to show an item if the `id, title, published` is not set?

gavin-mannings-wl commented on index.html:
```diff
@@ -65,52 +82,92 @@
-        let m=months[d.slice(5,7)-1];
-        let date=d.slice(8,10);
-        return(date+' '+m+' '+year);
+        if(d){
```
Dates are supplied from the API in a parseable format and don't need string manipulation.

```javascript
const d = new Date("2023-04-20T11:05:08Z")
console.log(d.toLocaleDateString('en-GB')); // '20/04/2023'
```

gavin-mannings-wl commented on index.html:
```diff
@@ -65,52 +82,92 @@
-    //XMLHttpRequest object is used to asynchronously communicate with the server and retrieve data. 
-    //Once the request is complete, the onreadystatechange event will be fired
+        else{
+          console.error('Error retrieving entries');
```
Should this be a console only error, or would you like to surface anything to the user?

gavin-mannings-wl commented:
I can see that you're still using bootstrap from the CDN, if you remove it, will your CSS need many amends?

NitulAU commented:
> I can see that you're still using bootstrap from the CDN, if you remove it, will your CSS need many amends?

 Oh you are correct. That is not required anymore👍🏻.

NitulAU committed:
Update server.js

NitulAU committed:
Add files via upload

public folder serve the static files(app.js & style.css) for express. Bringing out style.css allowed me to use sass language. This style.css is processed sass.

NitulAU committed:
Delete index.html

index.ejs file in view serves the html content.

NitulAU commented:
@gavin-mannings-wl please check the updates. I was confused how to integrate new changes to the same PR. What I did here is committed directly to the same branch. Don't know whether I messed something up. please check. Thanks in advance✌️.

NitulAU committed:
Add files via upload

sass architecture

NitulAU committed:
Update README.md

Added the process to execute the code.

gavin-mannings-wl commented on README.md:
````diff
@@ -5,4 +5,21 @@
+
+Prerequisites
+```
+npm install express
````
When you install dependencies with `npm`, if you use the `--save` flag it will create a `package.json` file that contains the versions of the plugins used to make the site work. The `package.json` should be committed as part of the repo

gavin-mannings-wl commented on sass/abstracts/_variables.scss:
```diff
@@ -0,0 +1,13 @@
+$col-black:#000;
+$col-header:#ff0000;
+$col-primary:#41aae6;
+$col-primary-bg:#41aae6d1;
```
Colours should be hex values, either 3 or 6 characters after a hash. https://www.pluralsight.com/blog/tutorials/understanding-hexadecimal-colors-simple

gavin-mannings-wl commented on sass/base/_base.scss:
```diff
@@ -0,0 +1,6 @@
+html{
+    position: absolute;
```
Not sure why you'd need this to be absolutely positioned? It's not normally recommended for the entire HTML document to be absolutely positioned

gavin-mannings-wl commented on sass/base/_typography.scss:
```diff
@@ -0,0 +1,5 @@
+a{
```
You may want to add a `:hover` pseudoclass so there is consistent styling when a cursor is placed over the link

gavin-mannings-wl commented on views/index.ejs:
```diff
@@ -0,0 +1,23 @@
+    <header class="header">
+      <button class="header__btn header__btn--page"><span>1</span>-<span>10</span></button>
+      <h1 class="header__title">Notices</h1> 
+      <button class="header__btn header__btn--next" role="navigation">next>></button>
```
The icons in `next >>` could be added via CSS `:after` pseudo element, which would keep the text free from clutter for screen readers

gavin-mannings-wl reviewed (approved):
I think this is fine to merge and move onto pagination, or any other tasks Manish has provided. There are some comments that can be picked up in the future but just being aware of them is fine.

NitulAU committed:
Update app.js

String manipulation inside function formatDate() function is removed.

NitulAU committed:
Update style.css

The text on button moved to CSS::after pseudo element.

NitulAU committed:
Update _notice.scss

passes alpha value through rgba() rather than appending in hex values.

NitulAU committed:
Update _header.scss

NitulAU committed:
Update _base.scss

NitulAU committed:
Update _variables.scss

standardized hex values.

NitulAU committed:
Update package.json

NitulAU committed:
Update _typography.scss

a :hover pseudo class added, so there is consistent styling when a cursor is placed over any link.

NitulAU commented:
> 

Is there any other way to define axios() in app.js other than this CDN method? I couldn't import  axios from node_modules folder because app.js is a browser based js code.

Since I'm using CDN here. I was not able to define axios() in test file for app.js.

gavin-mannings-wl commented:
> Is there any other way to define axios() in app.js other than this CDN method? I couldn't import axios from node_modules folder because app.js is a browser based js code.

I am assuming you are having problems including the code because node_modules is not made publicly available (correctly).

There are a number of ways to use axios in code without loading from a CDN. In your server you can add `axios.js` from node_modules to the `public` folder and load it like you do you app.js. This is fine in very small projects, however it makes it more difficult to manage dependencies. Loading it like this would add `axios` to the global scope and you can continue to call it as your existing code does.

If you copied the `axios.min.js` from node modules to your public folder (which exports using ESM syntax) then you can use the `type="module"` syntax to load the scripts in your HTML file:

```html
<script type="module" src="/axios.min.js"></script>
<script type="module" src="/app.js"></script>
```

If you updated your `app.js` script to import Axios using ESM syntax:

`import axios from './axios.min.js';`

Then this code will work correctly and you will have a import statement that should allow Jest to resolve the module too.

For larger projects it is useful to use a packager to concatenate dependencies to your application JS, or move the libraries needed from node_modules to the public directory. For this it is important to ESM module imports. We typically use Webpack in conjunction with Gulp to manage this. The basic webpack tutorials should allow you to do this and is essential on projects with thousands of lines and 10s-100s of different packages being used.
